datumaro/plugins/splitter.py

```diff
@@ -0,0 +1,480 @@
+                val = id_ratio[id_subsets.index(subset)]
+            else:
+                val = 0.0
+            trainval_ratio.append(val)
+        trainval_ratio = np.array(trainval_ratio)
+        total_ratio = np.sum(trainval_ratio)
+        if total_ratio < 1e-7:
+            trainval_splits = list(zip(["train", "val"], trainval_ratio))
+            log.warning(
+                "Sum of ratios is expected to be positive,\
+                got %s, which is %s" % (trainval_splits, total_ratio))
+        else:
+            trainval_ratio /= total_ratio # normalize
+            required = self._get_required(trainval_ratio)
+            for person_id, items in trainval.items():
+                np.random.shuffle(items)
+                by_attributes = self._group_by_attributes(items)
+                for attributes, indice in by_attributes.items():
+                    gname = 'person_id: {}, attributes: {}'.format(
+                        person_id, attributes)
+                    splits = self._split_indice(indice, gname,
+                        trainval_ratio, required)
+                    for subset, split in zip(trainval_subsets, splits):
+                        if len(split) > 0:
+                            by_splits[subset].extend(split)
+
+        for subset in self._subsets:
+            self._parts.append((set(by_splits[subset]), subset))
+
+    def get_subset_by_group(self, group:str):
+        assert group in self._group_map, \
+            "Unknown group '{}', available groups: {}".format(
+                group, self._group_map.keys()
+            )
+        group_id = self._group_map[group]
+        subset = self.select(lambda item: item.annotations[0].group == group_id)
+        return subset
+
+class SplitforDetection(TaskSpecificSplitter):
+    """
+    Splits dataset into train/val/test set for detection task.|n
+    For detection dataset, each image can have multiple bbox annotations.|n
+    Since one DataItem can't be included in multiple subsets at the same time,
+    the dataset can't be divided according to the bbox annotations.|n
+    Thus, we split dataset based on DatasetItem
+    while preserving label distribution as possible.|n
+    |n
+    Notes:|n
+    - Each DatsetItem is expected to have one or more Bbox annotations.|n
+    - Label annotations are ignored. We only focus on the Bbox annotations.|n
+    """
+
+    def __init__(self, dataset, splits, seed=None):
+        """
+        Parameters
+        ----------
+        dataset : Dataset
+        splits : list
+            A list of (subset(str), ratio(float))
+            Subset is expected to be one of ["train", "val", "test"].
+            The sum of ratios is expected to be 1.
+        seed : int, optional
+        """
+        super().__init__(dataset, seed)
+
+        subsets, sratio = self._validate_splits(splits)
+
+        ## 1. group by bbox label
+        by_labels = dict()
+        for idx, item in enumerate(self._extractor):
```

Comment:
Again, lets avoid iterating over input dataset, unless required.

Reply:
It is required. Can you suggest a better way to iterating over the input dataset?